src/Grpc.Net.Client/Internal/UserAgentGenerator.cs

```diff
@@ -0,0 +1,164 @@
+#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Runtime.Versioning;
+
+namespace Grpc.Net.Client.Internal
+{
+    internal static class UserAgentGenerator
+    {
+        internal static string GetUserAgentString()
+        {
+            var assembly = typeof(GrpcProtocolConstants).Assembly;
+
+            var assemblyVersion = assembly
+                .GetCustomAttributes<AssemblyInformationalVersionAttribute>()
+                .FirstOrDefault()?
+                .InformationalVersion;
+
+            var frameworkName = assembly
+                .GetCustomAttributes<TargetFrameworkAttribute>()
+                .FirstOrDefault()?
+                .FrameworkName;
+            
+            return GetUserAgentString(
+                processArch: RuntimeInformation.ProcessArchitecture,
+                runtimeVersion: IsMono() ? null : Environment.Version, 
+                assemblyVersion: assemblyVersion,
+                // RuntimeInformation.FrameworkDescription is only supported for
+                // .NET Framework 4.7.1 and above. If targetting net461 or earlier,
+                // the framework description will need to be resolved manually
+                // using reflection.
+                runtimeInformation: RuntimeInformation.FrameworkDescription,
+                frameworkName: frameworkName,
+                operatingSystem: GetOS());
+
+            bool IsMono() => Type.GetType ("Mono.Runtime") != null;
```

Comment:
```suggestion
            static bool IsMono() => Type.GetType("Mono.Runtime") != null;
```

Reply:
Is this the recommended way to test for mono?

Reply:
Seems to be. It's used extensively in the dotnet repos: https://dotnet.grep.app/search?q=IsMono&words=true